client/components/tinymce/plugins/wpcom-view/views/embed/index.js

```diff
@@ -1,140 +1,51 @@
-import actions from 'lib/embeds/actions';
 import EmbedView from './view';
+import getSiteEmbeds from 'state/selectors/get-site-embeds';
 import { getSelectedSiteId } from 'state/ui/selectors';
-import { SELECTED_SITE_SUBSCRIBE, SELECTED_SITE_UNSUBSCRIBE } from 'state/action-types';
+import { reduxDispatch, reduxGetState } from 'lib/redux-bridge';
+import { requestEmbed, requestEmbeds } from 'state/embeds/actions';
 
-export default class EmbedViewManager extends EventEmitter {
-	constructor() {
-		super();
+const EmbedViewManager = {};
 
-		this.sitesListener = this.updateSite.bind( this );
+EmbedViewManager.match = content => {
+	const siteId = getSelectedSiteId( reduxGetState() );
+	if ( ! siteId ) {
+		return;
 	}
 
-	onChange() {
-		this.emit( 'change' );
-		this.fetchSiteEmbeds();
-	}
+	reduxDispatch( requestEmbeds( siteId ) );
 
-	updateSite( selectedSiteId ) {
-		const siteId = selectedSiteId || getSelectedSiteId( this.store.getState() );
-
-		if ( ! this.hasOwnProperty( 'siteId' ) ) {
-			// First update (after adding initial listener) should trigger a
-			// fetch, but not emit a change event
-			this.siteId = siteId;
-			this.fetchSiteEmbeds();
-		} else if ( this.siteId !== siteId ) {
-			// Subsequent updates should neither emit a change nor trigger a
-			// fetch unless the site has changed
-			this.siteId = siteId;
-			this.onChange();
-		}
+	const embeds = getSiteEmbeds( reduxGetState(), siteId );
+	if ( ! embeds ) {
+		return;
 	}
 
-	addListener( event, listener, store ) {
-		super.addListener( event, listener );
-		if ( 'change' === event && 1 === this.listeners( event ).length ) {
-			this.store = store;
-			this.selectedSiteSubscriberId = uniqueId();
-			store.dispatch( {
-				type: SELECTED_SITE_SUBSCRIBE,
-				listener: this.sitesListener,
-			} );
-			this.embedsListListener = EmbedsListStore.addListener( this.onChange.bind( this ) );
-			this.embedsListener = EmbedsStore.addListener( this.onChange.bind( this ) );
-			this.updateSite();
-		}
-	}
-
-	removeListener( event, listener ) {
-		super.removeListener( event, listener );
-
-		if ( 'change' === event && ! this.listeners( event ).length ) {
-			this.store.dispatch( {
-				type: SELECTED_SITE_UNSUBSCRIBE,
-				listener: this.sitesListener,
-			} );
-			if ( this.embedsListListener ) {
-				this.embedsListListener.remove();
-			}
+	const rxLink = /(^|<p>)(https?:\/\/[^\s"]+?)(<\/p>\s*|$)/gi;
+	let currentMatch;
+	while ( ( currentMatch = rxLink.exec( content ) ) ) {
+		const url = currentMatch[ 2 ];
 
-			if ( this.embedsListener ) {
-				this.embedsListener.remove();
-			}
+		// Disregard URL if it's not a supported embed pattern for the site
+		const isMatchingPattern = embeds.some( pattern => pattern.test( url ) );
+		if ( ! isMatchingPattern ) {
+			continue;
 		}
-	}
 
-	fetchSiteEmbeds() {
-		if ( this.siteId && ! EmbedsListStore.get( this.siteId ) ) {
-			actions.fetch( this.siteId );
-		}
-	}
+		reduxDispatch( requestEmbed( siteId, url ) );
```

Comment:
The request to render the embed should be a part of the `EmbedView` component, using the classical query+selector pattern:
```jsx
function EmbedView() {
  return (
    <div>
      <QueryEmbed siteId={ ... } url={ ... } />
      { renderFrame() }
    </div>
  );
}

connect( state => ( {
  embed: getEmbed( state, siteId, url )
} ) );
```
The `match` function only needs the list of the site's embed regexps, and is not concerned with embed rendering at all.

Reply:
Good call, moved the actual embed render request to be handled by a query component that lives in `EmbedView`.